state-sync/src/coordinator.rs

```diff
@@ -1103,22 +1097,36 @@ impl<T: ExecutorProxyTrait> StateSyncCoordinator<T> {
             self.local_state.trusted_epoch()
         };
         self.local_state.verify_ledger_info(&response_li)?;
-        if let Some(li) = pending_li {
-            if li != response_li {
-                self.local_state.verify_ledger_info(&li)?;
+        self.validate_and_store_chunk(txn_list_with_proof, response_li.clone(), None)?;
+
+        // Need to sync with local storage to update synced version
+        self.sync_state_with_local_storage()?;
+        let synced_version = self.local_state.synced_version();
+
+        // Check if we've synced beyond our current highest known ledger info
+        if let Some(highest_known_ledger_info) = &self.highest_known_ledger_info {
+            if synced_version >= highest_known_ledger_info.ledger_info().version() {
+                self.highest_known_ledger_info = None;
             }
-            self.pending_ledger_infos.add_li(li);
         }
-        self.validate_and_store_chunk(txn_list_with_proof, response_li, None)?;
 
-        // need to sync with local storage to see whether response LI was actually committed
-        // and update pending_ledger_infos accordingly
-        self.sync_state_with_local_storage()?;
-        let new_version = self.local_state.synced_version();
+        // If we don't have a highest known ledger info, check if the new highest
+        // is appropriate for us.
+        if self.highest_known_ledger_info.is_none() {
+            if let Some(new_highest_li) = new_highest_li {
+                if new_highest_li != response_li {
+                    self.local_state.verify_ledger_info(&new_highest_li)?;
+                }
+
+                if synced_version < new_highest_li.ledger_info().version() {
+                    self.highest_known_ledger_info = Some(new_highest_li);
+                }
```

Comment:
Aah -- I ported this from existing code which assumes we *always* want to verify it before handling. One benefit of always doing this is that if the ledger info isn't valid, the peer should be penalized (as opposed to not penalizing the peer because we didn't end up needing the ledger info 😄).